### PR TITLE
Housekeeping:  Update Neuron training examples to SDK 2.28 and optimum-neuron 0.4.5

### PR DESCRIPTION
**Pull request by jimburtoft**
Just some housekeeping to make sure all the versions are current and working.  Still Trn1 only.  Trn2 support is forthcoming.

NxD examples (K8s + Slurm):
- Update DLC image to SDK 2.28.0 / PyTorch 2.9 / Python 3.12
- Pin neuronx-distributed clone to v2.27.0 tag for reproducible builds
- Update transformers to 4.48.0
- Fix missing code block close in Slurm README

Optimum-Neuron examples (K8s + Slurm):
- Rewrite train.py to use NeuronModelForCausalLM (replaces deprecated AutoModelForCausalLM + lazy_load_for_parallelism from ON 0.1.0)
- Use HfArgumentParser with NeuronTrainingArguments (replaces manual argparse)
- Update to optimum-neuron 0.4.5 with trl 0.24.0, peft 0.17.0
- Simplify K8s Dockerfile (single pip install of optimum-neuron[training])
- Update K8s DLC base to SDK 2.26.0 / PyTorch 2.8
- Update Slurm env to Python 3.10 with torch-neuronx 2.8
- Remove hardcoded --target=trn1 from NEURON_CC_FLAGS
- Replace tokenize_data K8s step with model download (NeuronSFTTrainer handles tokenization internally)
- Use consolidate_model_parallel_checkpoints_to_unified_checkpoint from optimum.neuron.models.training (replaces custom consolidation)
- Rewrite READMEs and K8s templates with updated CLI args

Tested on trn1.32xlarge with Llama 3.1 8B:
- NxD Slurm: 1 step, loss=12.58, 8 Neuron compilations passed
- NxD K8s: Docker build passed
- ON Slurm: 1 step, loss=2.21, MFU=19.04%, tokens/sec=34.88
- ON K8s: Docker build passed

## Changes

-

## Test Plan

**Environment:**
- AWS Service:
- Instance type:
- Number of nodes:

**Test commands:**
```bash

```

## Directory Structure

```
3.test_cases/
└── <framework>/                # e.g. pytorch, megatron, jax
    └── <library>/              # e.g. picotron, FSDP, megatron-lm
        └── <model>/            # e.g. SmolLM-1.7B (may be omitted for single-model cases)
            ├── Dockerfile      # Container / environment setup
            ├── README.md       # Overview, prerequisites, usage
            ├── slurm/          # Slurm-specific launch scripts
            ├── kubernetes/     # Kubernetes manifests
            └── hyperpod-eks/   # HyperPod EKS instructions
```

- Top-level files (`Dockerfile`, `README.md`, training scripts, configs) cover general setup.
- Subdirectories (`slurm/`, `kubernetes/`, `hyperpod-eks/`) contain service-specific launch instructions.
- Not all service subdirectories are required — include only the ones relevant to your test case.

## Checklist

- [ ] I have read the [contributing guidelines](https://github.com/awslabs/awsome-distributed-training/blob/main/CONTRIBUTING.md).
- [ ] I am working against the latest `main` branch.
- [ ] I have searched existing open and recently merged PRs to confirm this is not a duplicate.
- [ ] The contribution is self-contained with documentation and scripts.
- [ ] External dependencies are pinned to a specific version or tag (no `latest`).
- [ ] A README is included or updated with prerequisites, instructions, and known issues.
- [ ] New test cases follow the [expected directory structure](#directory-structure).

### CODE REVIEW

File: 3.test_cases/pytorch/optimum-neuron/llama3/slurm/fine-tuning/submit_jobs/0.create_env.sh [hunk 1]
```diff
@@ -8,19 +8,17 @@
...
+python3.10 -m venv /fsx/ubuntu/peft_ft/env_llama3_8B_peft
 source /fsx/ubuntu/peft_ft/env_llama3_8B_peft/bin/activate
 pip install -U pip
 
 python3 -m pip config set global.extra-index-url "https://pip.repos.neuron.amazonaws.com"
 
-python3 -m pip install --upgrade neuronx-cc==2.16.372.0 torch-neuronx==2.1.2.2.4.0 torchvision
-python3 -m pip install --upgrade neuronx-distributed==0.10.1 neuronx-distributed-training==1.1.1
+python3 -m pip install --upgrade neuronx-cc==2.* torch-neuronx==2.8.* torchvision
+python3 -m pip install --upgrade neuronx-distributed
```
Comment (KeitaW) [inline 3.test_cases/pytorch/optimum-neuron/llama3/slurm/fine-tuning/submit_jobs/0.create_env.sh]: ### Wildcard version pins

The previous version pinned exact package versions (`neuronx-cc==2.16.372.0`, `torch-neuronx==2.1.2.2.4.0`), but the update changes these to wildcard pins. The repo convention is to pin external dependencies to a specific version, tag, or commit — wildcards can produce different environments over time as new point releases are published, making issues harder to reproduce.

I'd suggest pinning to the exact versions tested — the Slurm README's `pip list` output for the neuronx-distributed example provides the exact versions for SDK 2.28. For the optimum-neuron example (SDK 2.26), the equivalent tested versions would be appropriate.

```suggestion
python3 -m pip install --upgrade neuronx-cc==2.22.12471.0 torch-neuronx==2.8.0.2.11.0 torchvision
python3 -m pip install --upgrade neuronx-distributed
```

File: 3.test_cases/pytorch/optimum-neuron/llama3/src/train.py [hunk 2]
```diff
@@ -1,119 +1,146 @@
...
+        },
+    )
+    model_final_path: str = field(
+        default="",
+        metadata={
+            "help": "Path to save the final model after training (in addition to output_dir checkpoints)."
+        },
+    )
+    lora_r: int = field(default=64, metadata={"help": "LoRA rank."})
+    lora_alpha: int = field(default=128, metadata={"help": "LoRA alpha."})
```
Comment (KeitaW) [inline 3.test_cases/pytorch/optimum-neuron/llama3/src/train.py]: ### LoRA default parameters don't match template values

The `ScriptArguments` dataclass declares `lora_r=64` and `lora_alpha=128` as defaults, but every K8s template and the Slurm `finetune-llama3-8B.sh` script passes `--lora_r=16 --lora_alpha=16`. Users who run `train.py` directly without explicit arguments (e.g., while experimenting) would get different LoRA behavior than what's documented and tested.

I'd suggest aligning the defaults with the values used throughout the PR:

```suggestion
    lora_r: int = field(default=16, metadata={"help": "LoRA rank."})
    lora_alpha: int = field(default=16, metadata={"help": "LoRA alpha."})
```

**Review by KeitaW — COMMENTED** [review]
## Review Batch 2/3 — Training Code Quality

The `train.py` rewrite is a big improvement — `NeuronModelForCausalLM` + `NeuronSFTTrainer` + `HfArgumentParser` is the right pattern. One default-value nit below.

File: 3.test_cases/pytorch/optimum-neuron/llama3/kubernetes/fine-tuning/Dockerfile [hunk 3]
```diff
@@ -1,37 +1,13 @@
-FROM 763104351884.dkr.ecr.us-east-1.amazonaws.com/huggingface-pytorch-training-neuronx:2.1.2-transformers4.43.2-neuronx-py310-sdk2.20.0-ubuntu20.04-v1.0
+FROM 763104351884.dkr.ecr.us-east-1.amazonaws.com/huggingface-pytorch-training-neuronx:2.8.0-transformers4.55.4-neuronx-py310-sdk2.26.0-ubuntu22.04
```
Comment (KeitaW) [inline 3.test_cases/pytorch/optimum-neuron/llama3/kubernetes/fine-tuning/Dockerfile:1]: ### Optimum-neuron DLC uses SDK 2.26, not 2.28

The PR title says "SDK 2.28" but this DLC image uses SDK 2.26.0. The neuronx-distributed example correctly uses SDK 2.28.0.

This is likely because the HuggingFace DLC hasn't been updated to SDK 2.28 yet. If that's the case, I'd suggest adding a brief note in the optimum-neuron K8s README (near the software versions table) clarifying that the DLC is on SDK 2.26 and will be updated when a 2.28-based DLC is available.

File: 3.test_cases/pytorch/optimum-neuron/llama3/slurm/fine-tuning/submit_jobs/0.create_env.sh (hunk 1 above)
Comment (jimburtoft) [reply 3.test_cases/pytorch/optimum-neuron/llama3/slurm/fine-tuning/submit_jobs/0.create_env.sh]: updated

File: 3.test_cases/pytorch/optimum-neuron/llama3/src/train.py (hunk 2 above)
Comment (jimburtoft) [reply 3.test_cases/pytorch/optimum-neuron/llama3/src/train.py]: changed to match template.  Retested and still works.

File: 3.test_cases/pytorch/optimum-neuron/llama3/kubernetes/fine-tuning/Dockerfile (hunk 3 above)
Comment (jimburtoft) [reply 3.test_cases/pytorch/optimum-neuron/llama3/kubernetes/fine-tuning/Dockerfile:1]: Added comment to README.

**Comment by jimburtoft** [comment]
@KeitaW All issues resolved.  Thanks for the great feedback!
